File: src/interchange/exceptions.py
"""Interchange-specific exceptions."""
from __future__ import annotations

from typing import Any


class InterchangeError(Exception):
    """Base class for all interchange errors."""


class FactConflictError(InterchangeError):
    """Raised when a baton contains facts that contradict WorldState.

    Attributes:
        conflicts: mapping of fact key -> (existing_value, proposed_value)
    """

    def __init__(self, conflicts: dict[str, tuple[Any, Any]]) -> None:
        self.conflicts = conflicts
        pairs = ", ".join(
            f"{k!r}: {old!r} != {new!r}" for k, (old, new) in conflicts.items()
        )
        super().__init__(f"Baton fact conflicts detected: {pairs}")


class CyclicDependencyError(InterchangeError):
    """Raised when the task graph contains a cycle."""

    def __init__(self, cycle: list[str]) -> None:
        self.cycle = cycle
        super().__init__(f"Cyclic dependency in task graph: {' -> '.join(cycle)}")


class RoutingError(InterchangeError):
    """Raised when no valid route can be determined for a task."""


class TranslationError(InterchangeError):
    """Raised when context translation between model formats fails."""


class TaskNotFoundError(InterchangeError):
    """Raised when a task_id is not found in the state backend."""

    def __init__(self, task_id: str) -> None:
        self.task_id = task_id
        super().__init__(f"Task not found: {task_id!r}")


class BatonValidationError(InterchangeError):
    """Raised when a baton fails structural validation (missing required fields)."""

File: packages/core/src/exceptions.ts
export class InterchangeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InterchangeError";
  }
}

export class FactConflictError extends InterchangeError {
  readonly conflicts: Record<string, [unknown, unknown]>;

  constructor(conflicts: Record<string, [unknown, unknown]>) {
    const detail = Object.entries(conflicts)
      .map(([k, [a, b]]) => `${k}: ${JSON.stringify(a)} vs ${JSON.stringify(b)}`)
      .join(", ");
    super(`Fact conflicts detected: ${detail}`);
    this.name = "FactConflictError";
    this.conflicts = conflicts;
  }
}

export class CyclicDependencyError extends InterchangeError {
  readonly cycle: string[];

  constructor(cycle: string[]) {
    super(`Cyclic dependency detected: ${cycle.join(" → ")}`);
    this.name = "CyclicDependencyError";
    this.cycle = cycle;
  }
}

export class RoutingError extends InterchangeError {
  constructor(message: string) {
    super(message);
    this.name = "RoutingError";
  }
}

export class TranslationError extends InterchangeError {
  constructor(message: string) {
    super(message);
    this.name = "TranslationError";
  }
}

export class TaskNotFoundError extends InterchangeError {
  readonly taskId: string;

  constructor(taskId: string) {
    super(`Task not found: ${taskId}`);
    this.name = "TaskNotFoundError";
    this.taskId = taskId;
  }
}

export class BatonValidationError extends InterchangeError {
  constructor(message: string) {
    super(message);
    this.name = "BatonValidationError";
  }
}
